docs/RFCS/sql_batching.md

```diff
@@ -0,0 +1,501 @@
+interacts with the KV layer, and it works quite well when the execution of a SQL statement
+needs to perform a few operations.
+
+However, the current semantics of `client.Txn` and `client.Batch` have one property that
+makes it suboptimal for this use case: only a single `client.Batch` can be executed at a
+time for a given `client.Txn`. This means that in order to run multiple statements concurrently,
+we would need to build up a single `client.Batch` between them and then run this batch when
+all statements are ready.
+
+Doing so would not be terribly difficult, but would be fairly intrusive into code surrounding
+the `sql.tableWriter` interface. The `sql.Executor` would need to inject some new object
+(`BatchCoalescer`?) that wraps a single `client.Batch`, packages multiple client requests
+from multiple goroutines into the batch, sends the batch only when all goroutines were ready,
+and then delegates the results out to the appropriate goroutines when the batch response arrives.
+During this implementation, it would almost certainly make sense to pull out a new interface
+in the `internal/client` package that incorporates the creation/execution of client requests
+(Get, Put, etc.), since these methods are already shared across `client.DB`, `client.Txn`, and
+`client.Batch`. Our new batch delegation object could then also implement this interface to
+allow for conditional indirection in the SQL layer.
+
+This approach gets more complicated when statements that issue multiple KV batches are
+considered. These cases are not rare, and arise for all statements that deal with foreign
+keys. This is also the case for UPSERT statements. The optimal solution here is not clear,
+but may reduce to the serial execution of these parts of the statement evaluation, or the
+decision to simply not parallelize any statement that requires more than one `client.Batch`
+execution at all.
+
+### Concurrent Batches
+
+An alternative approach looks to gain concurrency support at a layer below SQL, in KV logic.
+If the restriction that only one `client.Batch` can execute in a `client.Txn` at a time was
+lifted then SQL statement execution logic would not have to change between serially executed
+statements and concurrently executed statements. Each independent statement could run in its own
+goroutine and use it's own `client.Batch`
+
+This offers a number of benefits including that it is easier to reason about from the SQL
+layer, it eliminates the issue about statements that execute multiple batches, it is
+a generally useful improvement to the KV client API that could be beneficial elsewhere,
+and it removes the restriction of [coalesced batches](#coalesced-batches) that statements
+need to access the KV layer in lockstep. This last point means that this alternative should
+also perform better.
+
+However, it is unclear how difficult lifting this restriction would be. At the very least,
```

Comment:
My intuition is that this would be easier to implement. The KV layer has already been built with the assumption that batches can be split up by range and executed in arbitrary orders. The ratcheting of effects on the txn object (priority, orig/max/current/observed timestamps, and status) are designed to be commutative.

Certainly the last sentence is true; we'd need synchronization and changes to `TxnCoordSender`, but this seems tractable to me.